crates/sc-proof-of-time/src/pot_state.rs

```diff
@@ -0,0 +1,316 @@
+//! PoT state management.
+
+use crate::utils::LOG_TARGET;
+use crate::PotConfig;
+use parking_lot::Mutex;
+use sc_network::PeerId;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use subspace_core_primitives::{NonEmptyVec, PotKey, PotProof, PotSeed, SlotNumber};
+use subspace_proof_of_time::{PotVerificationError, ProofOfTime};
+use tracing::warn;
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum PotStateError {
+    #[error("Failed to extend chain: {expected}/{actual}")]
+    TipMismatch {
+        expected: SlotNumber,
+        actual: SlotNumber,
+    },
+
+    #[error("Proof for an older slot number: {tip_slot}/{proof_slot}")]
+    StaleProof {
+        tip_slot: SlotNumber,
+        proof_slot: SlotNumber,
+    },
+
+    #[error("Proof had an unexpected seed: {expected:?}/{actual:?}")]
+    InvalidSeed { expected: PotSeed, actual: PotSeed },
+
+    #[error("Proof had an unexpected key: {expected:?}/{actual:?}")]
+    InvalidKey { expected: PotKey, actual: PotKey },
+
+    #[error("Proof verification failed: {0:?}")]
+    InvalidProof(PotVerificationError),
+
+    #[error("Proof is too much into future: {tip_slot}/{proof_slot}")]
+    TooFuturistic {
+        tip_slot: SlotNumber,
+        proof_slot: SlotNumber,
+    },
+
+    #[error("Duplicate proof from peer: {0:?}")]
+    DuplicateProofFromPeer(PeerId),
+
+    #[error("Context mismatch of proof from peer: {0:?}")]
+    InvalidContextFromPeer(PeerId),
+}
+
+/// Action on extending the tip with a new proof.
+#[derive(Debug, Eq, PartialEq)]
+enum ExtendAction {
+    /// Try to merge pending proofs for future slots.
+    MergeFutureProofs,
+
+    /// Don't merge with future proofs.
+    NoMerge,
+}
+
+/// The shared PoT state.
+struct InternalState {
+    /// Last N entries of the PotChain, sorted by height.
+    chain: Vec<PotProof>,
+
+    /// Proofs for future slot numbers, indexed by slot number.
+    /// Each entry holds the proofs indexed by sender.
+    future_proofs: BTreeMap<SlotNumber, BTreeMap<PeerId, PotProof>>,
+}
+
+/// Wrapper to manage the state.
+struct StateManager {
+    /// Pot config
+    config: PotConfig,
+
+    /// PoT wrapper for verification.
+    proof_of_time: Arc<ProofOfTime>,
+
+    /// The PoT state
+    state: Mutex<InternalState>,
+}
+
+impl StateManager {
+    /// Creates the state.
+    pub fn new(config: PotConfig, proof_of_time: Arc<ProofOfTime>, chain: Vec<PotProof>) -> Self {
```

Comment:
I initially thought we may share the sync thread pool for parallel verification, so that we don't create too many threads. But saw your change to remove parallel verification, so now just a clone should be sufficient, changed it